### ISSUE TITLE
Bad performance for browsers

### ISSUE THREAD

**Issue by JeviScript** [justadudewhohacks/face-api.js#736]
I have found that performance is terribly slow for browsers. But that's the case at least if you use windows.
Tested right on the face-api site example : https://justadudewhohacks.github.io/face-api.js/face_recognition . I tested it on several different windows machines. It takes roughly 25 seconds to recognise faces.
https://user-images.githubusercontent.com/30739218/101359816-b224bf00-3894-11eb-801a-31e1c3bbfe4a.gif

On linux it works quite alright, taking 3 seconds with the example above. However there is one caveat. It works alright on chrome at this moment. For firefox there is a bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=1678652
https://bugzilla.mozilla.org/show_bug.cgi?id=1679671

In short, for firefox v83, there was a fix applied to how `failIfMajorPerformanceCaveat` property works. Now if a library uses `failIfMajorPerformanceCaveat: true` then webgl context creation fails and the library falls back to CPU, which is rather slow.
However, that's true for linux only for now. On windows it still uses webgl, and still rather slow (20-30 seconds) for any browser.

I am not sure if that's the issue of face-api.js or tensorflow itself

**Comment by cindyloo** [comment]
I had a similar issue (on a mac)
are you using tfjs2.0? that helps. I also am using a [webworker](https://github.com/justadudewhohacks/face-api.js/issues/47)
have a look here also-  it's pretty fast
[webcam example w react](https://justadudewhohacks.github.io/face-api.js/webcam_face_tracking/)

**Comment by JeviScript** [comment]
I am using face-api.js of 0.22.2 which in turn uses tfjs-core": "1.7.0".  I guess I can fork and upgrade tfjs to see if the issue is fixed there

**Comment by vladmandic** [comment]
@cindyloo  @JeviScript  
upgrade is not that trivial as there are breaking changes between tfjs 1.x and 2.x, including ops used inside models.  
take a look at <https://github.com/vladmandic/face-api>

**Comment by cindyloo** [comment]
@vladmandic I did read through your code but I was able to remove /nodemodules/faceapi tensorflow 1.7 so it would default to tfjs2.0 and it worked fine...

**Comment by vladmandic** [comment]
@cindyloo I've heard that being done before, but not sure what is it supposed to accomplish since TFJS 1.7 is bundled inside face-`api.js` during build process. And if you try to rebuild `face-api.js` using TFJS 2.x, it fails.  
(`face-api.js` incorrectly lists `@tensorflow/tfjs-core": "1.7.0` as dependency when in reality its a devDepencency)

**Comment by cindyloo** [comment]
hmm - if I look inside the plainscript face-api.js, I don't see any specific bundling. Wouldn't that mean it would look to the package settings and node_modules to determine the library?

**Comment by vladmandic** [comment]
face-api.js includes bundled tfjs-core.

If you don't even look at the code, look at sizes - face-api on its own is ~200kb, remaining ~600kb is tfjs-core.

**Comment by cindyloo** [comment]
thanks for the clarification @vladmandic. I can see it invoking the tflib now

so, having revisited the codebase, I am getting reasonably good performance. The fix for me was the throttling of the face detection and using an async call for the analysis
'analyzeVideo' is essentially the onPlay method in the example

eg:
```
const pullAndPostFrameTimer = (videoEl, mode) => {

  const faceDetectorOptions = getFaceDetectorOptions();

  if (!videoEl || videoEl.ended || !faceDetectorOptions || faceDetectorOptions == undefined) {
    setTimeout(function() {
      pullAndPostFrameTimer(videoEl, mode);
    }, THROTTLETIMER);
    return;
  }
  let canvasVideoBuff = document.querySelector(
    `#videoBuff${mode}`
  ) as HTMLCanvasElement;

  analyzeVideo(videoEl, mode, faceDetectorOptions);
 
  setTimeout(function() {
    pullAndPostFrameTimer(videoEl, mode);
  }, THROTTLETIMER);
};
```

**Comment by vladmandic** [comment]
i'm not sure i understand as this example is not complete.
is `analyzeVideo` async function? if yes, then you have race scenario that is reduced a bit by having a delay between frames.

why not put a promise in `analyzeVideo` and then do something like:
```js
analyzeVideo(videoEl, mode, faceDetectorOptions).then(() => requestAnimationFrame(() => pullAndPostFrameTimer(videoEl, mode));
```
or easier to read with async/await
```js
await analyzeVideo(videoEl, mode, faceDetectorOptions);
requestAnimationFrame(() => pullAndPostFrameTimer(videoEl, mode));
```
that way loop will run as fast as possible and there is no race.

**Comment by cindyloo** [comment]
@vladmandic so appreciate the note to race conditions. I had to review those within javascript ([this is a good review](https://medium.com/@slavik57/async-race-conditions-in-javascript-526f6ed80665)) as well as  `requestAnimationFrame`

so here's the thing- performance seems to suffer a little with using `requestAnimationFrame` as there is no delay imposed and it attempts w every result to detect a face...I guess I could do a sleep or something?

**Comment by vladmandic** [comment]
When it comes to video, I don't like adding sleep to solve problems.  
Ideally, you could make it a two separate loops, both using requestAnimationFrame:  
- One loop that handles rendering and uses last known results  
- Second loop that handles processing  
  Ideally get ImageData on main thread and pass it as a reference to web worker thread   
  to handle detection and send results back to main thread.
  That way main UI thread gets minimal blocking and first loop will run at full refresh rate of the screen.  
  (that is one benefit of requestAnimationFrame - it will run up to refresh rate and not faster if not needed)

**Comment by cindyloo** [comment]
I took @vladmandic's advice to separate the rendering logic from the facial analysis (see the analyzeVideo vs. drawFaces routines). I alos tried to make my code work with `requestAnimationFrame` as well as with `requestVideoFrameCallback` but it ended up being really laggy (not sure what I was doing incorreclty.  The best solution I have found is to use the setTimeout/setInterval pattern eg:
```
const pullAndPostFrameTimer = (mode, anonymousParticipants) => {
  const faceDetectorOptions = getFaceDetectorOptions();

  // for each element tag passed in, add a timer
 participants.map((p, index) => {
    const videoEl = whichVideo(`${mode}Video${p}`) as HTMLVideoElement;

    if (
      !videoEl ||
      videoEl.ended ||
      videoEl.readyState < 3 ||
      !faceDetectorOptions ||
      faceDetectorOptions == undefined
    ) {
      return false;
    }

    analyzeVideo(videoEl, faceDetectorOptions); // async
    drawFaces(mode, p);  //async
    return true;
  });

  setTimeout(function () {
    pullAndPostFrameTimer(mode, participants);
  }, THROTTLETIMER); // 8000 while waiting for faces, then speed it up to 2000ms (my best reaction time is between 1000-2000ms)
};

const FaceVisualiser = (params) => {
  const { mode, participants } = params;
  useEffect(() => {
    if (participants) {
      if (pullAndPostFrameTimer(mode, participants)) {
        THROTTLETIMER = FACE_DETECTING; // 2000 ms
        return;
      }
    }
    THROTTLETIMER = NO_FACE_YET;
    return function cleanup() {
      THROTTLETIMER = NO_FACE_YET;
    };
  }),
    [participants];
  return <div />;
};
```
